database/migrations/2022_02_02_203008_filesize_size_variants.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+
+/**
+ * Add a filesize column for size variants to make it
+ * easier to implement filesystems with a Lychee backend
+ * that needs to know in advance the size of files, for example.
+ *
+ * The size of the original photo is also moved to the original size variant.
+ */
+class FilesizeSizeVariants extends Migration
+{
+	private const VAR_TAB = 'size_variants';
+	private const PHOTO_FK = 'photo_id';
+	private const TYPE_COL = 'type';
+	private const WIDTH_COL = 'width';
+	private const HEIGHT_COL = 'height';
+	private const ID_COL = 'id';
+	private const TYPE_ORIGINAL = 0;
+	private const PHOTOS_TAB = 'photos';
+	private const SIZE_COL = 'filesize';
+
+	/**
+	 * Run the migrations.
+	 *
+	 * @return void
+	 */
+	public function up()
+	{
+		// To avoid doing I/O on every photo, which would be prohibitive
+		// on large instances, and because JPEG compression makes approximations
+		// from original filesize too approximative, just set filesize to -1
+		// for every variant except the original, indicating that the size
+		// has not been calculated yes.
+		// The real calculation can be done by calling `artisan lychee:variant_filesize` from the CLI.
+		// This will replace the default value by the actual filesize but takes time.
+		Schema::table(self::VAR_TAB, function (Blueprint $table) {
+			$table->bigInteger(self::SIZE_COL)->nullable(false)->default(-1);
```

Comment:
I know that we need some kind of "sentinel" element which indicates a missing size, because we don't add the correct size during migration.

However, I wonder whether

1. we should use `null` instead of `-1`,
2. make the column unsigned instead, and
3. enhance the console command `lychee:variant_filesize` to make the column not-nullable after all file sizes have been added

This will give us the chance to have a correctly typed column in the future. This is meant as an open discussion, because I have no string opinion about that.

However, I have an opinion about not calculating file sizes right during migration. Have we actually tested how long it takes? I have the feeling that we only claim that it will be inefficient. I did some preliminary tests some time ago and reported that on Gitter. Even for a large DB like @ildyria's it should only be around 10minutes. Yes, this is longer than normally, but IMHO not prohibitive. I would really like to see @ildyria, @d7415 and @kamil4 to run the console command on their installation for a test and report some runtime figures.

If the results turn out to be not too bad, I would really recommend to calculate file sizes during migration. That would allow us to make the column a non-nullable, unsigned integer. Otherwise we will be hunted either by a signed or nullable column in the future.

Reply:
> it should only be around 10minutes. Yes, this is longer than normally, but IMHO not prohibitive.

10 mins for a field that 99% of users won't care about? That sounds prohibitive.

Reply:
@nagmat84 I cannot be really helpful on this one. `-1` seems fine to me and modifying the schema in a command seems odd, intuitively.   At first I used `0` as a default value and I changed it because if a filesize was really `0` (should not, but technically can be), I was not sure that the paging query would correctly work, but `lazyById` seems to do the job.

What do you think ? Could `0` do the job as a default value, avoiding nullable or signed column ?

Reply:
IMHO, in this case `0` should do as a sentinel for "no value" as well, because there should be no real file with size `0`.

Reply:
Done again in 790bf82124.

Reply:
Looks good. :+1: Conversation can be resolved!